### DISCUSSION TITLE
Keybindings don't work until I've switched to the R terminal window and back

### DISCUSSION THREAD

**Discussion by lokesh-krishna** [R-nvim/R.nvim#85] [Q&A] [answered]
I have auto-start enabled so whenever I open a Quarto document or R script, I get the R terminal window as well. However, none of the R.nvim keybindings work until I've manually switched to the R window once and then back. Is this expected behavior or am I missing something?

**Comment by jalvesaq** [comment]
With lazy.nvim and lazy=false on Linux, the keybindings are immediately
available.

>

**Comment by lokesh-krishna** [comment]
I'm using lazy too and do have `lazy=false` set but the bindings just don't take effect until I've been to the R terminal once. 

```lua
return {
  'R-nvim/R.nvim',
  dependencies = {
    {
      'R-nvim/cmp-r',
      config = function()
        require('cmp_r').setup {
          fun_data_1 = {
            'select',
            'rename',
            'mutate',
            'filter',
            'relocate',
            'summarise',
          },
        }
      end,
    },
  },
  lazy = false,
  config = function()
    local opts = {
      auto_start = 'always',
      active_window_warn = false,
    }
    require('r').setup(opts)
    vim.g.rout_follow_colorscheme = true

    -- Keybinding to view R tables using DT
    local function show_table()
      local node = vim.treesitter.get_node { ignore_injections = false }
      local text = vim.treesitter.get_node_text(node, 0)
      local cmd = [[call slime#send("DT::datatable(]] .. text .. [[)" . "\r")]]
      vim.cmd(cmd)
    end

    vim.keymap.set('n', '<leader>dt', show_table, { desc = '[D]ata [T]able' })

    -- Keybinding to insert pipe
    vim.keymap.set('i', '<m-m>', ' |>')
  end,
}
```

This is my config for R.nvim. Is there something I'm doing wrong?

**Comment by jalvesaq** [comment]
I replaced my config with yours, and it works as expected: the key bindings are immediately active. I don't have `slime` installed, though.

**Comment by lokesh-krishna** [comment]
I tried again with Slime uninstalled and still can't get it to work without switching over to the R window. I am guessing it must be something elsewhere in my config that's causing the issue then?

**Comment by lokesh-krishna** [comment]
Update! The keybindings do work! They just don't show up with which-key until I've switched to the R window once. I have to hit `<localleader>` and follow it quickly with the keys for even a slight delay seems to do their usual action instead. 

`<localleader>pp` to send paragraph works if I do it quickly enough, else it just pastes whatever's in the clipboard twice. 

If I've switched to the R window once though, it actually brings up the which-key panel after I've pressed `<localleader>`. 

![image](https://github.com/R-nvim/R.nvim/assets/20235646/8305c852-72ac-4cd3-b9ab-b11c2ac2d8c0)

Is this the behaviour that you too are describing?

**Comment by jalvesaq** [comment]
You have to press the next letter: `r` to see `s` (Send to R: summary), etc.

**Comment by jalvesaq** [comment]
`+prefix` means: there is more than one key binding starting with this letter.

**Reply by lokesh-krishna** [reply]
Yes yes, that bit I do get. I was referring to this panel only showing up if I've visited the R window once.

**Comment by jalvesaq** [comment]
I get the panel before starting R and immediately after starting R. No need to move the cursor.

**Comment by jalvesaq** [comment]
I noted that `RConfigShow` doesn't get autocompletion immediately after starting Neovim. I have to do an editing action such as going to Insert mode and back to Normal mode to make autocompletion work.

**Comment by srearl** [comment]
I am seeing similarly odd behaviour but that does not seem to be fixed by context switching:
[nvim-r.txt](https://github.com/R-nvim/R.nvim/files/14627700/nvim-r.txt)

![nvim_r](https://github.com/R-nvim/R.nvim/assets/8332564/8cbd699e-44de-4f54-a133-cfe04eb0f045)

BTW @lokesh-krishna , thanks for this helpful bit: `vim.keymap.set("i", "<leader>s", " |>")`

**Comment by PMassicotte** [comment]
@srearl I think you are using an old version. For example, there are only 3 keymaps for sending function:

```lua
    -- Function
    create_maps("nvi", "RSendAllFun",    "fa", "<Cmd>lua require('r.send').funs(0, true, false)")
    create_maps("nvi", "RSendCurrentFun",   "fc", "<Cmd>lua require('r.send').funs(0, false, false)")
    create_maps("nvi", "RDSendCurrentFun",   "fd", "<Cmd>lua require('r.send').funs(0, false, true)")
```

**Reply by srearl** [reply]
Ah, right @PMassicotte, I have been referring to the old documentation. Good catch! However, the short cuts for executing selected text (i.e., `\s*`) appear to be the same in the [new docs](https://github.com/jamespeapen/Nvim-R/wiki/Use) and so are still not executing as expected.

**Comment by jalvesaq** [comment]
The link for the "new docs" refers to Nvim-R, not R.nvim.

**Reply by srearl** [reply]
:confused: gaa, so sorry for my confusion, that is the below link from [this](https://github.com/R-nvim/R.nvim) page, which seemed logical but I should have paid closer attention:

![doc_link](https://github.com/R-nvim/R.nvim/assets/8332564/18db4bec-aa30-4cc0-b90a-13fe65d5667d)

**Comment by jalvesaq** [comment]
@srearl please look at https://github.com/R-nvim/R.nvim#transitioning-from-nvim-r

**Reply by srearl** [reply]
...but, still, `\ss` carried forward and is not behaving as expected: it works in normal mode (similarly to sending a paragraph) but not in visual mode where it substitutes the selected text for the letter 's'.

**Reply by jalvesaq** [reply]
I can replicate the bug, but only because I have this in my config:

```lua
vim.api.nvim_buf_set_keymap(0, "v", "<Enter>", "<Plug>RSendSelection", {})
```

That is, the bug is in the output of `:RMapsDesc` because `<LocalLeader>ss` still is the key binding to send the previous selection in Normal mode, but not for sending selection in Visual mode.

**Reply by srearl** [reply]
ah, gotcha', I was using your configuration as a guide, and executing a visual selection works as expected if I comment out the following lines:

```lua
hook = {
  after_config = function ()
    if vim.o.syntax ~= "rbrowser" then
      vim.api.nvim_buf_set_keymap(0, "n", "<Enter>", "<Plug>RDSendLine", {})
      vim.api.nvim_buf_set_keymap(0, "v", "<Enter>", "<Plug>RSendSelection", {})
    end
  end,
},
```
:+1:

**Comment by jalvesaq** [comment]
The reference is wrong in the README! I will fix it. Thank you for reporting!

**Comment by lokesh-krishna** [comment]
Somehow this has worsened to which-key panel not showing up even on context switching and the keybinds only working if I do them very rapidly. Would really appreciate any help in troubleshooting this.

**Comment by PMassicotte** [comment, marked as answer]
What happens if you present the leader key before. This should do the trick.

**Reply by lokesh-krishna** [reply]
This did it! If you don't mind then could you explain what might be going on here? Does it not have anything to do with my R.nvim setup at all then?

**Comment by PMassicotte** [comment]
That is a good question, I have noted that I needed to do it also. I will investigate and report if I find something.

**Comment by lokesh-krishna** [comment]
Upgraded to which-key 3.5 and this issue seems to have fixed itself. Not sure how but keybinds now work right off the bat whenever I open a R file.

**Comment by jalvesaq** [comment]
Thanks for the feedback!
